First pass at program group reorganization.

cmnbroad opened the pull request:
Program group definitions only, including Picard placeholders. Based on @sooheelee 's spreadsheet.

sooheelee commented:
Some of the categories we are removing are still in the HelpConstants, e.g. DOC_CAT_SPARK*.

Here's that spreadsheet again. It's the fourth tab `categories_summaries`: https://docs.google.com/a/broadinstitute.org/spreadsheets/d/19SvP6DHyXewm8Cd47WsM3NUku_czP2rkh4L_6fd-Nac/edit?usp=sharing

lbergelson commented on src/main/java/org/broadinstitute/hellbender/cmdline/programgroups/AlignmentDuplicateBQSRProgramGroup.java:
```diff
@@ -0,0 +1,15 @@
+import org.broadinstitute.barclay.argparser.CommandLineProgramGroup;
+import org.broadinstitute.hellbender.utils.help.HelpConstants;
+
+public class AlignmentDuplicateBQSRProgramGroup  implements CommandLineProgramGroup {
```
This is a confusing group.  Maybe "BamPreprocessing" would be less specific but more clear?

sooheelee replied:
Sure. Geraldine mentioned the Picard folks may not like the current category name so BAM Preprocessing will likely cause fewer protests.

lbergelson commented on src/main/java/org/broadinstitute/hellbender/cmdline/programgroups/MetaGenomicsProgramGroup.java:
```diff
@@ -0,0 +1,13 @@
+import org.broadinstitute.barclay.argparser.CommandLineProgramGroup;
+import org.broadinstitute.hellbender.utils.help.HelpConstants;
+
+public class MetaGenomicsProgramGroup  implements CommandLineProgramGroup {
```
Would PathogenDiscoveryProgramGroup be clearer?

sooheelee replied:
Metagenomics, with lower-case `g` please.

lbergelson commented on src/main/java/org/broadinstitute/hellbender/cmdline/programgroups/ReadDataProgramGroup.java:
```diff
@@ -0,0 +1,15 @@
+import org.broadinstitute.barclay.argparser.CommandLineProgramGroup;
+import org.broadinstitute.hellbender.utils.help.HelpConstants;
+
+public class ReadDataProgramGroup implements CommandLineProgramGroup {
```
I don't understand what this group is, the description makes it sound like it should be `IntervalProgramGroup`

"Tools that process genomic intervals in various formats"

cmnbroad replied:
Thats totally a mistake...

sooheelee replied:
Can we add an `s` to make _IntervalsProgramGroup_ please.

lbergelson commented on src/main/java/org/broadinstitute/hellbender/utils/help/HelpConstants.java:
```diff
@@ -19,62 +19,99 @@ public static String forumPost(String post) {
+    public final static String DOC_CAT_EXAMPLE = "Example Tools";
+    public final static String DOC_CAT_EXAMPLE_SUMMARY = "Example tools that show developers how to implement new tools";
+
+    public final static String DOC_CAT_METAGENOMICS = "MetaGenomics";
```
seems like this would be clearer as PathogenDetection instead of MetaGenomics, but I could be wrong about that.

ldgauthier replied:
"pathogen" is very judgemental -- there are a lot of applications of PathSeq that are not necessarily disease-related

lbergelson replied:
I wouldn't want to offend any beneficial microbes!

sooheelee replied:
The workflow tool names start with _Pathseq_ so folks will immediately know what Metagenomics refers to. @mwalker174 requested we use _metagenomics_ instead of _pathogen_.

magicDGS commented:
A general idea: it might be useful to have a developer javadoc for indicating the purpose of the program group. That will clarify purposes for things that are not user-related.

cmnbroad commented:
Meta-comments for reviewers: the new program groups in this PR are based on @sooheelee 's spreadsheet, including some that are placeholders for things that will soon live in Picard, but aren't accessible from there yet.

I've left the old program groups intact because they're still being referenced by tools. As the doc PRs are merged in, eventually these will be left dangling with no references, and then we'll remove them. In the meantime we'll have some redundancies (ReadProgramGroup will be replaced by ReadDataProgramGroup, or whatever we settle on).

cmnbroad commented:
@sooheelee ReadProgramGroup and VariantProgramGroup should not have been marked "Picard placeholder". Those are existing PGs currently in use in GATK, but destined to be removed. I'll remove the placeholder comments from them.

BaseCalling isn't included here because based on the spreadsheet it looks to be Picard only, so it will be added there. It can be referenced in the future in GATK, but I didn't add a a placeholder for it since the placeholder would be unused.

sooheelee commented:
I've minor requests of adding letters etc. Otherwise, good to go on my part. @vdauwera This one is a priority.

sooheelee commented:
One additional thing to do is to update the fourth tab of the spreadsheet so folks know which program groups to apply to their tools.

cmnbroad commented:
Responded to review comments. Since we now have a mix of "obsolete", "approved", and "placeholder" program groups that are easily confused (some obsolete groups look similar to new approved groups, i.e., ReadProgramGroup which is obsolete, and ReadsDataGroup, which is approved, but a placeholder for Picard. So I added a comment to each PG class marking it as obsolete, or as a placeholder. For the new, approved groups, I just added a comment containing the summary text.

I also fixed an existing problem where we QCProgramGroup was used for some actual tools, and some tools that were jut for testing. I updated the test tools to point to the TestProgramGroup, and obsoleted QCProgramGroup (like all of the obsolete groups, there are still tools referencing it).

The sooner we merge this the better.

vdauwera commented on src/main/java/org/broadinstitute/hellbender/cmdline/programgroups/DiagnosticsAndQCProgramGroup.java:
```diff
@@ -0,0 +1,18 @@
+import org.broadinstitute.hellbender.utils.help.HelpConstants;
+
+/**
+ * Tools that collect sequencing quality related and comparative metrics.
```
sequencing quality-related

cmnbroad replied:
Done.

vdauwera commented on src/main/java/org/broadinstitute/hellbender/cmdline/programgroups/MetagenomicsProgramGroup.java:
```diff
@@ -3,9 +3,14 @@
 
-public final class FastaProgramGroup implements CommandLineProgramGroup {
+/**
+ * Tools that detect pathogens, e.g. bacteria.
```
Description should be more general

-> Tools that perform metagenomic analysis, e.g. microbial community composition and pathogen detection

cmnbroad replied:
Done.

cmnbroad replied:
Done.

vdauwera commented on src/main/java/org/broadinstitute/hellbender/cmdline/programgroups/OtherProgramGroup.java:
```diff
@@ -0,0 +1,18 @@
+import org.broadinstitute.hellbender.utils.help.HelpConstants;
+
+/**
+ * Miscellaneous tools, e.g. those that aide in data streaming.
```
aide->aid

cmnbroad replied:
Done.

vdauwera commented on src/main/java/org/broadinstitute/hellbender/cmdline/programgroups/ShortVariantDiscoveryProgramGroup.java:
```diff
@@ -0,0 +1,17 @@
+import org.broadinstitute.hellbender.utils.help.HelpConstants;
+
+/**
+ * Tools that reassemble reads and call short variants. Tools also call non-reference probability, genotype and
```
This description seems overly specific about internal mechanisms of the callers; would replace with e.g.:

Tools that perform variant calling and genotyping for short variants (SNPs, SNVs and Indels)

cmnbroad replied:
Done.

vdauwera commented on src/main/java/org/broadinstitute/hellbender/cmdline/programgroups/VariantEvaluationProgramGroup.java:
```diff
@@ -0,0 +1,19 @@
+import org.broadinstitute.hellbender.utils.help.HelpConstants;
+
+/**
+ * Tools that evaluate and refine variant calls, e.g. with annotations not offered by the engine".
```
"annotations not offered by the engine" is super cryptic outside of the GATK context so will get pushback in Picard; let's rephrase to  e.g.:

Tools that evaluate and refine variant calls, e.g. by adding annotations that are not calculated during variant calling

(I'm not crazy about this formulation either but it's more explicit at least)

cmnbroad replied:
Done.

vdauwera commented on src/main/java/org/broadinstitute/hellbender/cmdline/programgroups/VariantFilteringProgramGroup.java:
```diff
@@ -0,0 +1,19 @@
+import org.broadinstitute.hellbender.utils.help.HelpConstants;
+
+/**
+ * Tools that filter variants by annotating the FILTER column.
```
I'm on the fence about "by annotating the FILTER column" -- it seems overly specific and mechanism-based. I think it might hurt more than it helps. "Tools that filter variants" should be sufficient.

cmnbroad replied:
Done.
